### PR TITLE
Fixed using of Deflater when compressing PDF.

### PR DESCRIPTION
**Pull request by mdudek**

### CODE REVIEW

**Comment by Uzlopak** [comment]
I will probably adapt this code for jspdf. But what is it fixing?

**Comment by mdudek** [comment]
> I will probably adapt this code for jspdf. But what is it fixing?

These commits fixed processing of larger pdf document with PDF compression:

- there was not correctly used Deflater class to compress PDF content
- fixed usage of String.fromCharCode for conversion of array buffer to string for larger data. There was thrown 'Maximum call stack size exceeded exception'.

**Comment by Uzlopak** [comment]
I try to applay the changes into jsPDF

https://github.com/MrRio/jsPDF/pull/2215

Do you know why the pdf validator [here](https://www.pdf-online.com/osa/validate.aspx) shows that the FlateEncode stream  is invalid?

**Comment by mdudek** [comment]
> I try to applay the changes into jsPDF
> 
> [MrRio#2215](https://github.com/MrRio/jsPDF/pull/2215)
> 
> Do you know why the pdf validator [here](https://www.pdf-online.com/osa/validate.aspx) shows that the FlateEncode stream is invalid?

Download veraPDF from here: http://verapdf.org/software/

This tool gives you detailed info/report about validation issues.

File: dist/jspdf.debug.js [hunk 1]
```diff
@@ -1130,7 +1130,9 @@
             arr.set(new Uint8Array([120, 156]));
             arr.set(p, 2);
             arr.set(new Uint8Array([adler32 & 0xff, adler32 >> 8 & 0xff, adler32 >> 16 & 0xff, adler32 >> 24 & 0xff]), p.byteLength + 2);
-            p = String.fromCharCode.apply(null, arr);
+            p = arr.reduce(function (data, byte) {
+              return data + String.fromCharCode(byte);
```
Comment (yGuy) [inline dist/jspdf.debug.js]: For a better performance (GC pressure), we should try to use String.fromCharCode with more than a single argument as much as possible. E.g. we could split "arr" into multiple 1024 byte arrays, convert these once and then join all.

Comment (mdudek) [reply dist/jspdf.debug.js]: Done. ;-)

File: dist/jspdf.debug.js [hunk 2]
```diff
@@ -1130,9 +1130,14 @@
...
             arr.set(p, 2);
             arr.set(new Uint8Array([adler32 & 0xff, adler32 >> 8 & 0xff, adler32 >> 16 & 0xff, adler32 >> 24 & 0xff]), p.byteLength + 2);
-            p = arr.reduce(function (data, byte) {
-              return data + String.fromCharCode(byte);
-            }, '');
+            var strings = [],
+                chunkSize = 0xffff; // There is a maximum stack size. We cannot call String.fromCharCode with as many arguments as we want
+
+            for (var j = 0; j * chunkSize < arr.length; j++) {
+              strings.push(String.fromCharCode.apply(null, arr.subarray(j * chunkSize, (j + 1) * chunkSize)));
```
Comment (yGuy) [inline dist/jspdf.debug.js:1137]: Cool - thank you! 

Some thoughts:

We could be slightly less aggressive here - I haven't tested and I don't know how browsers will react, here. IMHO "maximum stack size" should not only be affected by the arguments to the last method, but also to the calling stack - if the calling stack is deep, (which we have no control over) could this become a problem again? Haven't tested, but I guess we could also just reduce the chunksize to 0x8fff and hope for the best ;-)   (for reference: https://stackoverflow.com/a/22747272/351836 )



I would have added a comment to explain that the upper bound of the subarray function is automatically clamped to valid values. I was not aware of this feature and thought that the code will fail in some cases.  

// end is clamped to valid array indices by "subarray"

Comment (mdudek) [reply dist/jspdf.debug.js:1137]: I have tried latest version of Chrome/Firefox/Edge and 0xffff works. I will test Safari on wednesday. For sure we can reduce amount of arguments to 0x8fff, if it's more safe...

**Review by HackbrettXXX — APPROVED** [review]

**Pull request merged**